### PR TITLE
[Feat][Attention] Add GQA paged prefill FP8 KV cache op

### PR DESCRIPTION
**Pull request by superAngGao**
## Summary

Add a storage-only FP8 KV cache variant for paged GQA prefill:

- Introduce `GroupedQueryAttentionPrefillPagedWithFP8KVCacheFwdOp`.
- Keep `q`, `k_new`, and `v_new` in the attention dtype for the current chunk.
- Store paged KV cache as `float8_e4m3fn`.
- Dequantize old cache pages online with scalar `k_scale` / `v_scale` tensors of shape `(1,)`.
- Quantize the current chunk into FP8 E4M3 when appending to the paged cache.
- Add a spec-driven manifest entry and targeted tests.

This is the storage-only path. Fused RoPE + FP8 append and FP8 tensor-core attention compute are intentionally left as follow-up kernel variants.

## Tests

Ran with the nightly test image `tileops-runner:nightly-tl019-fullstack-no-tileops-ldfix` on H200:

```bash
python -m pytest -q tests/ops/attention/test_gqa_prefill_paged.py -k fp8 --tb=short -s
# 3 passed, 17 deselected

python -m pytest -q tests/ops/attention/test_gqa_prefill_paged.py --tb=short -s
# 20 passed

python scripts/validate_manifest.py --check-op GroupedQueryAttentionPrefillPagedWithFP8KVCacheFwdOp
# All manifest checks passed, with expected shape/dtype parity and bench integration warnings.
```

### CODE REVIEW

**Review by Gabbering — COMMENTED** [review]
## goose goose review — `570a2242`

*honk.* A clean structural copy-paste from the non-FP8 paged kernel with FP8 quant/dequant bolted on. Mostly correct. Mostly. One real bug, one perf concern, one test gap.

### goose Bugs
- `tileops/kernels/attention/gqa_fwd.py:2199` — The `page_size % block_n == 0` fast-path for reading old cache pages assumes the entire `block_n` tile sits inside a single logical page. That's only true when `tile_start` is page-aligned **or** `page_offset + block_n <= page_size`. The code only checks `page_size % block_n == 0`, then computes `physical_start` from `tile_start`'s page. If `tile_start` happens to be mid-page and `block_n == page_size`, it reads past the physical page boundary into an adjacent (wrong) physical page. The non-FP8 sibling kernel at line ~1858 has the **exact same latent bug**, so this was faithfully copied. However: for the default config (`block_n=64, page_size=64`) the condition `page_size % block_n == 0` is true and `tile_end <= old_len` guarantees `tile_start` is a multiple of `block_n` (== `page_size`), so it happens to be safe for `page_size == block_n`. It breaks if someone autotunes to `block_n=32` with `page_size=64` and the old_len isn't block_n-aligned — `tile_start` can be 32, page_idx would be 0, page_offset 32, and you'd read indices 32..63 from that physical page, which is correct. Actually wait — `tile_start = k_idx * block_n`, so it's always a multiple of `block_n`, and since `page_size % block_n == 0`, `page_offset + block_n <= page_size` always holds. The goose retracts its fangs on this one. Carry on.

- `tileops/kernels/attention/gqa_fwd.py:2159` (FP8 append, `page_size % block_n` fast path) — There is **no synchronisation barrier** between the FP8 cache append (guarded by `by < heads_kv`) and the attention read loop that reads from `k_pages`/`v_pages`. All `heads` CTAs in the same `(bx, bz)` group share the same `k_pages`/`v_pages` global memory. CTAs with `by >= heads_kv` skip the append and fall straight into the attention loop, where the boundary tile (`tile_start < old_len < tile_end` or `tile_start >= old_len`) reads `k_new`/`v_new` directly — so those CTAs don't actually read the freshly-written FP8 data for the *current* chunk. The old-cache tiles (`tile_end <= old_len`) only read pages written by previous calls, so there's no WAR hazard there either. The non-FP8 sibling has exactly the same structure and ships; so this is architecturally consistent and not a new bug introduced here. Fine.

  Actually — let me look harder. For the **old cache dequant path** the kernel reads `k_pages[..., cur_kv_head, ...]` which is data written by a *previous* kernel launch. No hazard. For the **new chunk path** (`tile_start >= old_len`) it reads from `k_new`/`v_new` (read-only inputs). For the **boundary path** it conditionally reads either old cache or `k_new`/`v_new`. So the freshly-appended FP8 data in `k_pages` is never consumed within the same launch. The goose is satisfied. No sync bug.

  **However**, there IS a subtle data-race in the append itself: multiple CTAs with the same `(bz, by)` but **different `bx`** values all perform the append for their own `block_m` tile of new tokens, which is fine — they write disjoint positions. But within a single CTA, if `by < heads_kv` and `by != cur_kv_head` (which is `by // groups`), the append writes to head `by` while the attention reads from head `cur_kv_head`. These are different head indices when `groups > 1`, so different memory locations. No race. Good.

  The goose found no actual bugs in the kernel logic. Disappointing but correct.

### feather Performance
- `tileops/kernels/attention/gqa_fwd.py:2148-2210` — Every element in the old-cache dequant path does **two casts and a multiply per element, per KV**: `fp8 → float32 → multiply → cast to dtype`. For the `page_size % block_n == 0` fast-path, this is `2 * block_n * dim` casts + `block_n * dim` multiplies, repeated for both K and V. The scale values `k_scale[0]` and `v_scale[0]` are loaded from global memory inside the innermost parallel loop — relying on the compiler to hoist them to registers. If tilelang doesn't hoist scalar loads out of `T.Parallel` (it usually does, but worth verifying), every thread re-reads from global. Consider loading `k_scale[0]` and `v_scale[0]` into a local variable before the pipelined loop. This is a nit on a brand-new kernel path, but for long-prefix scenarios (e.g. the 32k-prefix workload in the manifest) the dequant overhead on old pages dominates, so it matters.

### egg Test gaps
- `tests/ops/attention/test_gqa_prefill_paged.py:213-312` — The FP8 test only exercises `dtype=float16`. The manifest declares `q: {dtype: "float16 | bfloat16"}`, and the Op's `_validate_attention_dtype` accepts bfloat16, but there is zero test coverage for `bfloat16 + float8_e4m3fn` cache. The existing non-FP8 tests parametrize over both dtypes. Add at least one `bfloat16` case — FP8 dequant → bf16 cast has different rounding behaviour than → fp16, and the tolerance may need adjusting.

- `tests/ops/attention/test_gqa_prefill_paged.py:213-312` — All three FP8 test cases use `page_size=64, block_n=64` (default config). The non-FP8 tests have a dedicated `test_gqa_prefill_paged_with_kv_cache_page_sizes` that exercises `page_size ∈ {16, 32, 128}` to hit the `block_n % page_size == 0` and generic fallback code paths in the kernel. The FP8 kernel has identical branching for those paths (with added dequant), but none of them are tested. A `page_size=16` or `page_size=128` case would exercise the `block_n % page_size` and per-element fallback dequant paths respectively.

File: tileops/kernels/attention/gqa_fwd.py [hunk 1]
```diff
@@ -2109,6 +2110,392 @@ def forward(self, q: torch.Tensor, k_new: torch.Tensor, v_new: torch.Tensor,
...
+                            page_idx = logical_pos >> T.int32(page_size_log2)
+                            page_offset = logical_pos - page_idx * page_size
+                            if new_pos < q_len:
+                                physical_pos = block_table[bz, page_idx] * page_size + page_offset
+                                k_pages[physical_pos, by, d] = T.clamp(
+                                    T.Cast("float32", k_new[q_start + new_pos, by, d]) /
+                                    k_scale[0], fp8_min, fp8_max)
+                                v_pages[physical_pos, by, d] = T.clamp(
+                                    T.Cast("float32", v_new[q_start + new_pos, by, d]) /
+                                    v_scale[0], fp8_min, fp8_max)
```
[the remark this answers is not in the document]
Comment (superAngGao) [reply tileops/kernels/attention/gqa_fwd.py]: Done in 79dd506. I added a `quantize_fp8` `T.macro` and switched all FP8 append branches to use it, so the clamp/cast/scale expression stays in one place.

File: tileops/kernels/attention/gqa_fwd.py [hunk 2]
```diff
@@ -2109,6 +2110,392 @@ def forward(self, q: torch.Tensor, k_new: torch.Tensor, v_new: torch.Tensor,
...
+                        acc_s,
+                        transpose_B=True,
+                        policy=T.GemmWarpPolicy.FullRow)
+                    if use_softcap:
+                        apply_softcap(acc_s)
+                    online_softmax(acc_s, scores_max, scores_max_prev, scores_scale,
+                                   scores_sum, logsum)
+                    T.copy(acc_s, acc_s_cast)
+                    rescale(acc_o, scores_scale)
+                    T.gemm(acc_s_cast, v_shared, acc_o, policy=T.GemmWarpPolicy.FullRow)
```
[the remark this answers is not in the document]
Comment (superAngGao) [reply tileops/kernels/attention/gqa_fwd.py:2355]: Done in 79dd506. I added a matching `dequantize_fp8` `T.macro` and replaced the old-cache dequantization sites across the page-size fast paths and fallback path.

**Comment by superAngGao** [comment]
Updated in 79dd506.

Changes:
- Extracted FP8 append quantization into `quantize_fp8` T.macro.
- Extracted FP8 old-cache dequantization into `dequantize_fp8` T.macro.
- Expanded FP8 tests to cover bf16 attention dtype and page_size 16/128 in addition to the existing page_size 64 paths.

Validation:
- `python3 -m ruff check tileops/kernels/attention/gqa_fwd.py tests/ops/attention/test_gqa_prefill_paged.py`
- `PYTHONPATH=. python3 scripts/validate_manifest.py --check-op GroupedQueryAttentionPrefillPagedWithFP8KVCacheFwdOp`
- `python3 -m pytest -q tests/ops/attention/test_gqa_prefill_paged.py -k fp8 --tb=short -s` -> 6 passed
- `python3 -m pytest -q tests/ops/attention/test_gqa_prefill_paged.py --tb=short -s` -> 23 passed

**Review by Gabbering — COMMENTED** [review]
goose *goose skimmed `79dd506` — nothing to honk about.*

File: tileops/ops/attention/gqa.py [hunk 3]
```diff
@@ -944,6 +950,238 @@ def total_memory(self) -> int:
...
+                f"got {tuple(block_table.shape)}")
+
+        for name, tensor in [("q", q), ("k_new", k_new), ("v_new", v_new)]:
+            if tensor.dtype != self.dtype:
+                raise ValueError(f"Expected {name}.dtype {self.dtype}, got {tensor.dtype}")
+        for name, tensor in [("k_pages", k_pages), ("v_pages", v_pages)]:
+            if tensor.dtype != self.cache_dtype:
+                raise ValueError(
+                    f"Expected {name}.dtype {self.cache_dtype}, got {tensor.dtype}")
+        for name, tensor in [("k_scale", k_scale), ("v_scale", v_scale)]:
```
Comment (Ibuki-wind) [inline tileops/ops/attention/gqa.py:1117]: GroupedQueryAttentionPrefillPagedWithFP8KVCacheFwdOp._validate_forward_inputs only checks k_scale/v_scale shape and dtype while the kernel divides by those values in quantize_fp8; zero or non-finite scales will silently produce saturated/NaN FP8 cache and output -> reject non-finite or <= 0 k_scale/v_scale values before launching the kernel, with a focused validation test.

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
### Overall
Two blocking issues remain in the changed API surface; address the inline comments before the next round.

File: tileops/ops/attention/gqa.py (hunk 3 above)
Comment (superAngGao) [reply tileops/ops/attention/gqa.py:1117]: Done in ee558b9. `_validate_forward_inputs` now rejects non-finite or non-positive `k_scale` / `v_scale` values before launching the kernel, after the dtype check. I also added a focused smoke test covering zero, negative, inf, and nan scale inputs.

**Comment by superAngGao** [comment]
Updated in ee558b9.\n\nChanges:\n- Reject non-finite or non-positive `k_scale` / `v_scale` in `GroupedQueryAttentionPrefillPagedWithFP8KVCacheFwdOp._validate_forward_inputs` before launching the kernel.\n- Added focused smoke coverage for zero, negative, inf, and nan FP8 KV cache scales.\n\nValidation:\n- `python3 -m ruff check tileops/ops/attention/gqa.py tests/ops/attention/test_gqa_prefill_paged.py`\n- `PYTHONPATH=. python3 scripts/validate_manifest.py --check-op GroupedQueryAttentionPrefillPagedWithFP8KVCacheFwdOp`\n- `python3 -m pytest -q tests/ops/attention/test_gqa_prefill_paged.py -k invalid_scales --tb=short -s` -> 4 passed\n- `python3 -m pytest -q tests/ops/attention/test_gqa_prefill_paged.py -k fp8 --tb=short -s` -> 10 passed

**Comment by superAngGao** [comment]
Follow-up a31a440 fixes the validate-manifest failure from the stricter ctor-default check: `cache_dtype` now defaults to the manifest dtype token `"float8_e4m3fn"` and is normalized to `torch.float8_e4m3fn` inside the constructor, while still accepting callers that pass the torch dtype explicitly.\n\nRevalidated locally with the same focused ruff/manifest/invalid-scale test command.

**Review by Gabbering — COMMENTED** [review]
goose *goose skimmed `a31a440` — nothing to honk about.*

**Comment by stelladuyx** [comment]
Read through the full diff and the existing review thread. A few observations after the latest round of updates:

**What's in good shape**

- The `quantize_fp8` / `dequantize_fp8` T.macro extraction (79dd506) is the right call — it removes the inline cast duplication and makes the intent clear.
- Scale validation (ee558b9) is thorough: shape, dtype, device, contiguity, finiteness, and positivity are all checked before the kernel launches.
- The `cache_dtype` default normalization (a31a440) correctly handles both the string token and the torch dtype, which is consistent with how the manifest and the constructor interact.
- Test coverage now includes bf16, page_size 16/128, and the invalid-scale rejection cases — the gaps Gabbering flagged are addressed.

**One remaining concern: `k_scale[0]` / `v_scale[0]` inside the pipelined loop**

In `_gqa_prefill_paged_with_fp8_kv_cache_fwd_main`, `k_scale[0]` and `v_scale[0]` are accessed inside `T.Parallel` blocks that sit inside `T.Pipelined`. Whether TileLang hoists these scalar loads to registers depends on the backend — if it doesn't, every thread in every pipeline stage re-reads from global memory. For the long-prefix workload in the manifest (32k prefix, 270k physical tokens), the dequant loop dominates and this could matter. Worth either loading them into a local scalar before the pipelined loop, or confirming the compiler hoists them.

**Minor**

- The `_select_gqa_prefill_paged_with_fp8_kv_cache_fwd_kernel_cls()` function is a one-liner that always returns the same class. That's fine as a dispatch hook for future variants, but it's worth a short comment so the next reader doesn't wonder why it exists.
- The roofline entry reuses `gqa_prefill_with_kv_cache_fwd_roofline` from the non-FP8 op. That formula presumably doesn't account for the dequant FLOP cost on old pages. Fine for now if it's understood to be approximate, but worth noting in the YAML comment.

Overall the implementation is clean and the blocking issues from the previous round have been addressed. The scale-hoisting question is the only thing I'd want confirmed before merge.

**Comment by superAngGao** [comment]
Follow-up pushed in de26b67.

Changes:
- Hoisted `k_scale[0]` / `v_scale[0]` into local TileLang scalars before the FP8 append/dequant paths, so the pipelined old-cache loop no longer references the global scale tensors directly.
- Added a short selector-hook comment for the FP8 paged-cache kernel dispatch helper.
- Documented that the FP8 manifest roofline reuses the non-FP8 paged-prefill estimate and does not model old-cache dequantization separately.

Validation on nightly Docker with host GPU 1 (`GPU-bea8d0b6-e3c4-082c-e524-50f7791c9a1e`):
- `python3 -m ruff check tileops/kernels/attention/gqa_fwd.py tileops/ops/attention/gqa.py tests/ops/attention/test_gqa_prefill_paged.py`
- `python3 scripts/validate_manifest.py --check-op GroupedQueryAttentionPrefillPagedWithFP8KVCacheFwdOp` -> passed with existing parity/bench warnings
- `python3 -m pytest -q tests/ops/attention/test_gqa_prefill_paged.py -k "fp8 or invalid_scales" --tb=short -s -o cache_dir=/tmp/pytest-cache` -> 10 passed, 17 deselected

**Review by Gabbering — COMMENTED** [review]
goose *goose skimmed `de26b67` — nothing to honk about.*

**Review by michaelwithu — APPROVED** [review]
Adds `GroupedQueryAttentionPrefillPagedWithFP8KVCacheFwdOp` — a storage-only FP8 KV cache variant for paged GQA prefill. The kernel dequantizes old cache pages online with scalar `k_scale`/`v_scale` and appends the current chunk quantized into FP8 E4M3 pages. Includes kernel, Op, manifest entry, and comprehensive tests (causal/non-causal, softcap, fp16/bf16, page_size 16/64/128, invalid scale rejection).

LGTM.

**Review by stelladuyx — APPROVED** [review]
Reviewed the full diff and the follow-up commits. The implementation is solid:

- FP8 append and dequantize paths are correct — the `quantize_fp8`/`dequantize_fp8` macros centralize the cast/scale logic cleanly, and the three-way page-boundary dispatch (aligned page, multi-page block_n, general) handles all the edge cases.
- The `k_scale`/`v_scale` validation (finite + positive) was the right fix — silent NaN/saturation in the FP8 cache would have been very hard to debug.
- `local_k_scale`/`local_v_scale` hoisted out of the inner loop in the latest commit — good catch.
- Test coverage is reasonable: 6 parametrized smoke cases covering causal/non-causal, softcap, bf16, and multiple page sizes, plus the scale validation tests.

One minor note for a follow-up: the `roofline` comment acknowledges it doesn't model the dequantization cost. That's fine for now, but worth revisiting when benchmarking against the non-FP8 path to avoid misleading efficiency numbers.

Approving.

**Review by Ibuki-wind — APPROVED** [review]
